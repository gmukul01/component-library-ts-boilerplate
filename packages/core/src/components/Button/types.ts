import { HTMLProps, WithThemeProp } from '@core-utils/types';

export interface Props extends HTMLProps<HTMLButtonElement>, WithThemeProp {
    type?: 'button' | 'reset' | 'submit';
    solid?: boolean;
    flat?: boolean;
    outlined?: boolean;
}
