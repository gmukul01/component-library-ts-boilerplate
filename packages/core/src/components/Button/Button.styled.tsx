import { css, styled } from '@core-styled';
import { ButtonTheme, defaultTheme } from '@react-basis/theme';
import { Props } from './types';

const outlined = (theme: ButtonTheme) => css`
    border: 1px solid ${theme.outlined.borderColor};
    color: ${theme.outlined.textColor};
`;

const solid = (theme: ButtonTheme) => css`
    background-color: ${theme.solid.bgColor};
    color: ${theme.solid.textColor};
`;

const flat = (theme: ButtonTheme) => css`
    color: ${theme.flat.textColor};
`;

export const ButtonStyled = styled('button')<Props>`
    background-color: transparent;
    border: none;
    padding: 10px;
    margin: ${({ theme }) => theme.spacing.S};
    border-radius: 3px;
    user-select: none;

    &:hover {
        cursor: pointer;
    }

    &:focus {
        outline: none;
    }

    ${props => props.flat && flat(props.theme.button)};
    ${props => props.outlined && outlined(props.theme.button)};
    ${props => props.solid && solid(props.theme.button)};
`;

ButtonStyled.defaultProps = {
    theme: defaultTheme
};
