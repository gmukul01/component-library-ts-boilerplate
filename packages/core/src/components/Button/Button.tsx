import { WithStyle } from '@utils/types';
import React from 'react';
import Text from '../Text';
import { ButtonStyled } from './Button.styled';
import { Props } from './types';

import { isValidStringOrNumber } from '@utils/helpers/ReactHelper';

const Button: React.FunctionComponent<Props> & WithStyle = props => {
    return (
        <ButtonStyled {...props}>
            {React.Children.map(props.children, c => {
                return isValidStringOrNumber(c) ? (
                    <Text textWeight="Strong" uppercase>
                        {c}
                    </Text>
                ) : (
                    c
                );
            })}
        </ButtonStyled>
    );
};
Button.displayName = 'Button';
Button.Style = ButtonStyled;

export default Button;
