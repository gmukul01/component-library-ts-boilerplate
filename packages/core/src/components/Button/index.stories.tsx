import { wInfo } from '@storybook-utils';
import { select } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Button from './Button';

const buttonOptions = ['solid', 'flat', 'outlined'];
storiesOf('Components', module).add(
    'Button',
    () => (
        <Button
            {...{
                solid: 'solid' === select('Button Option', buttonOptions, 'solid'),
                flat: 'flat' === select('Button Option', buttonOptions, 'solid'),
                outlined: 'outlined' === select('Button Option', buttonOptions, 'solid')
            }}
        >
            Demo Button
        </Button>
    ),
    wInfo('Button Component with below props')
);
