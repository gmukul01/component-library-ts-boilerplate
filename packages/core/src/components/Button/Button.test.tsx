import { render } from '@core-test-utils';
import 'jest-styled-components';
import React from 'react';
import Button from './Button';

describe('Button component', () => {
    it('should render properly with flat prop', () => {
        const { container } = render(<Button flat>Flat Button</Button>);
        expect(container).toMatchSnapshot();
    });
    it('should render properly with outlined prop', () => {
        const { container } = render(<Button outlined>Outlined Button</Button>);
        expect(container).toMatchSnapshot();
    });
    it('should render properly with solid prop', () => {
        const { container } = render(<Button solid>Solid Button</Button>);
        expect(container).toMatchSnapshot();
    });

    it('should wrap children into Text component when rendering string or number', () => {
        const { container } = render(
            <Button solid>
                Some text
                <i>some icon</i>
                <div>element2</div>
            </Button>
        );
        expect(container).toMatchSnapshot();
    });
});
