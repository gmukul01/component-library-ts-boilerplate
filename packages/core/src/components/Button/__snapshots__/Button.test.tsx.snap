// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render properly with flat prop 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 700;
  color: inherit;
  text-transform: uppercase;
}

.c0 {
  background-color: transparent;
  border: none;
  padding: 10px;
  margin: 8px;
  border-radius: 3px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #5392f9;
}

.c0:hover {
  cursor: pointer;
}

.c0:focus {
  outline: none;
}

<div>
  <button
    class="c0"
  >
    <strong
      class="c1"
    >
      Flat Button
    </strong>
  </button>
</div>
`;

exports[`Button component should render properly with outlined prop 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 700;
  color: inherit;
  text-transform: uppercase;
}

.c0 {
  background-color: transparent;
  border: none;
  padding: 10px;
  margin: 8px;
  border-radius: 3px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid #5392f9;
  color: #5392f9;
}

.c0:hover {
  cursor: pointer;
}

.c0:focus {
  outline: none;
}

<div>
  <button
    class="c0"
  >
    <strong
      class="c1"
    >
      Outlined Button
    </strong>
  </button>
</div>
`;

exports[`Button component should render properly with solid prop 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 700;
  color: inherit;
  text-transform: uppercase;
}

.c0 {
  background-color: transparent;
  border: none;
  padding: 10px;
  margin: 8px;
  border-radius: 3px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #5392f9;
  color: #ffffff;
}

.c0:hover {
  cursor: pointer;
}

.c0:focus {
  outline: none;
}

<div>
  <button
    class="c0"
  >
    <strong
      class="c1"
    >
      Solid Button
    </strong>
  </button>
</div>
`;

exports[`Button component should wrap children into Text component when rendering string or number 1`] = `
.c1 {
  font-size: 14px;
  font-weight: 700;
  color: inherit;
  text-transform: uppercase;
}

.c0 {
  background-color: transparent;
  border: none;
  padding: 10px;
  margin: 8px;
  border-radius: 3px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: #5392f9;
  color: #ffffff;
}

.c0:hover {
  cursor: pointer;
}

.c0:focus {
  outline: none;
}

<div>
  <button
    class="c0"
  >
    <strong
      class="c1"
    >
      Some text
    </strong>
    <i>
      some icon
    </i>
    <div>
      element2
    </div>
  </button>
</div>
`;
